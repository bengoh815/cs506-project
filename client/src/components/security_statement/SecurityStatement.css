#security-statement-overlay {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0, 0, 0, 0.5);
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 1000;
}

#security-statement-overlay div {
  padding: 2rem;
  background: white;
  border-radius: 5px;
  display: flex;
  flex-direction: column;
  align-items: center;
  box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
}

#security-statement-heading {
  margin-bottom: 2rem;
}

#security-statement-body {
  text-align: left;
  padding-left: 2rem;
  padding-right: 2rem;
}
