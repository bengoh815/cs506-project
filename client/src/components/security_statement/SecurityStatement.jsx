/******************************************************************************
 * Filename: SecurityStatements.jsx
 * Purpose:  An overlay component that displays security statement to user.
 * Author:   Victor Nguyen & Don Ma
 *
 * Description:
 * This file contains the SecurityStatement component that displays a security
 * statement to the user. The component is designed to be displayed as an overlay
 * on top of the main application content and is hidden by default. The user can
 * click on the overlay to close it and return to the main application content.
 *
 * Usage:
 * To use this component, simply import it into the desired file and render it.
 *
 ******************************************************************************/

import React, { useState, useEffect } from "react";
import "./SecurityStatement.css";

const SecurityStatement = () => {
  // State to store the visibility of the security statement
  const [isVisible, setIsVisible] = useState(true);

  // Add an event listener to close the popup when the user clicks outside of it
  useEffect(() => {
    const closePopup = () => setIsVisible(false);

    if (isVisible) {
      window.addEventListener("click", closePopup);
    }

    return () => window.removeEventListener("click", closePopup);
  }, [isVisible]);

  if (!isVisible) return null;

  return (
    <div
      id="security-statement-overlay"
      data-testid="security-statement-overlay" // Test id for testing
    >
      <div>
        <h1
          id="security-statement-heading"
          data-testid="security-statement-heading"
        >
          Security Statement
        </h1>
        <p id="security-statement-body" data-testid="security-statement-body">
          • Our system will only record audio in between when the user has
          clicked the start and stop recording button. <br />
          <br />
          • Our system will only store the MIDI file converted from audio
          recordings. <br />
          <br />
          • The system will state that the audio recordings will only be used
          for converting into the MIDI file. <br />
          <br />
          • MIDI files will be stored in the database and open for all users to
          see. <br />
          <br />
          • The system will not save audio recordings because users may want to
          avoid having their voice being recorded and stored on the system.
          <br />
          <br />• Our system will be vulnerable to denial-of-service attacks.
        </p>
      </div>
    </div>
  );
};

export default SecurityStatement;
