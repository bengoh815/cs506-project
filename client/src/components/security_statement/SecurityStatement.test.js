/******************************************************************************
 * Filename: SecurityStatements.test.js
 * Purpose:  Tests the SecurityStatement component.
 * Author:   Victor Nguyen
 *
 * Description:
 * This file contains tests for the SecurityStatement component. The tests
 * verify that the component renders correctly and that the security statement
 * is displayed to the user. The tests also verify that the security statement
 * can be closed by clicking outside of it.
 *
 * Usage:
 * Run the tests using the command `npm test`.
 *
 ******************************************************************************/

import { act, render, screen, cleanup } from "@testing-library/react";
import SecurityStatement from "./SecurityStatement";

describe("Security Statement Component", () => {
  // Cleans up the DOM after each test to ensure a clean environment.
  afterEach(() => {
    cleanup();
  });

  it("whole security statement component renders", () => {
    // Arrange - Render the component
    render(<SecurityStatement />);
    const securityStatementElement = screen.getByTestId(
      "security-statement-overlay",
    );

    // Assert
    expect(securityStatementElement).toBeInTheDocument();
  });

  it("security statement heading renders", () => {
    // Arrange
    render(<SecurityStatement />);
    const securityStatementHeading = screen.getByTestId(
      "security-statement-heading",
    );

    // Assert
    expect(securityStatementHeading).toBeInTheDocument();
  });

  it("correct security statement heading content", () => {
    // Arrange
    render(<SecurityStatement />);
    const securityStatementHeading = screen.getByTestId(
      "security-statement-heading",
    );

    // Assert
    expect(securityStatementHeading).toHaveTextContent("Security Statement");
  });

  it("security statement body renders", () => {
    // Arrange
    render(<SecurityStatement />);
    const securityStatementBody = screen.getByTestId("security-statement-body");

    // Assert
    expect(securityStatementBody).toBeInTheDocument();
  });

  it("correct security statement body content", () => {
    // Arrange
    render(<SecurityStatement />);
    const securityStatments = [
      "Our system will only record audio in between when the user has clicked the start and stop recording button.",
      "Our system will only store the MIDI file converted from audio recordings.",
      "The system will state that the audio recordings will only be used for converting into the MIDI file.",
      "The system will state that the audio recordings will only be used for converting into the MIDI file.",
      "MIDI files will be stored in the database and open for all users to see.",
      "The system will not save audio recordings because users may want to avoid having their voice being recorded and stored on the system.",
      "Our system will be vulnerable to denial-of-service attacks.",
    ];
    const securityStatementBody = screen.getByTestId("security-statement-body");

    // Assert
    securityStatments.forEach((statement) => {
      expect(securityStatementBody).toHaveTextContent(statement);
    });
  });

  it("security statement closes when document is clicked", () => {
    // Arrange
    render(<SecurityStatement />); // Render the component
    const securityStatementElement = screen.getByTestId(
      "security-statement-overlay",
    );

    // Must wrap code that causes React state updates in act().
    act(() => {
      window.dispatchEvent(new MouseEvent("click")); // Simulates click event on window
    });

    // Assert
    expect(securityStatementElement).not.toBeInTheDocument();
  });
});
