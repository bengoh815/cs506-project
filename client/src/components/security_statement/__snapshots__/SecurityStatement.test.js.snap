// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  data-testid="security-statement-overlay"
  id="security-statement-overlay"
>
  <div>
    <h1
      data-testid="security-statement-heading"
      id="security-statement-heading"
    >
      Security Statement
    </h1>
    <p
      data-testid="security-statement-body"
      id="security-statement-body"
    >
      • Our system will only record audio in between when the user has clicked the start and stop recording button. 
      <br />
      <br />
      • Our system will only store the MIDI file converted from audio recordings. 
      <br />
      <br />
      • The system will state that the audio recordings will only be used for converting into the MIDI file. 
      <br />
      <br />
      • MIDI files will be stored in the database and open for all users to see. 
      <br />
      <br />
      • The system will not save audio recordings because users may want to avoid having their voice being recorded and stored on the system.
      <br />
      <br />
      • Our system will be vulnerable to denial-of-service attacks.
    </p>
  </div>
</div>
`;
